File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJClassInheritanceIndex.java
package org.cappuccino_project.ide.intellij.plugin.indices;

import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.psi.stubs.StubIndexKey;
import org.cappuccino_project.ide.intellij.plugin.exceptions.IndexNotReadyRuntimeException;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;


public class ObjJClassInheritanceIndex extends ObjJStringStubIndexBase<ObjJClassDeclarationElement> {
    private static final ObjJClassInheritanceIndex INSTANCE = new ObjJClassInheritanceIndex();
    private static final StubIndexKey<String, ObjJClassDeclarationElement> KEY = IndexKeyUtil.createIndexKey(ObjJClassInheritanceIndex.class);
    private static final int VERSION = 0;
    private ObjJClassInheritanceIndex() {}

    public static ObjJClassInheritanceIndex getInstance() {
        return INSTANCE;
    }

    @Override
    public int getVersion() {
        return super.getVersion() + VERSION;
    }

    @NotNull
    @Override
    protected Class<ObjJClassDeclarationElement> getIndexedElementClass() {
        return ObjJClassDeclarationElement.class;
    }

    @NotNull
    @Override
    public StubIndexKey<String, ObjJClassDeclarationElement> getKey() {
        return KEY;
    }

    @NotNull
    public List<String> getChildClassesAsStrings(@NotNull String parentClassName, @NotNull Project project) {
        return getChildClassesRecursive(new ArrayList<>(), parentClassName, project);
    }

    private List<String> getChildClassesRecursive(@NotNull final List<String> descendants, @NotNull final String className, @NotNull Project project) {
        if (DumbService.isDumb(project)) {
            throw new IndexNotReadyRuntimeException();
        }
        for (ObjJClassDeclarationElement classDeclarationElement : get(className, project)) {
            String currentClassName = classDeclarationElement.getClassNameString();
            if (descendants.contains(currentClassName)) {
                continue;
            }
            descendants.add(currentClassName);
            getChildClassesRecursive(descendants, currentClassName, project);
        }
        return descendants;
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJClassInheritanceIndex.kt
package org.cappuccino_project.ide.intellij.plugin.indices

import com.intellij.openapi.project.DumbService
import com.intellij.openapi.project.Project
import com.intellij.psi.stubs.StubIndexKey
import org.cappuccino_project.ide.intellij.plugin.exceptions.IndexNotReadyRuntimeException
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement

import java.util.ArrayList


class ObjJClassInheritanceIndex private constructor() : ObjJStringStubIndexBase<ObjJClassDeclarationElement<*>>() {

    protected override val indexedElementClass: Class<ObjJClassDeclarationElement<*>>
        get() = ObjJClassDeclarationElement<*>::class.java

    override fun getVersion(): Int {
        return super.getVersion() + ObjJIndexService.INDEX_VERSION + VERSION
    }

    override fun getKey(): StubIndexKey<String, ObjJClassDeclarationElement<*>> {
        return KEY
    }

    fun getChildClassesAsStrings(parentClassName: String, project: Project): List<String> {
        return getChildClassesRecursive(ArrayList(), parentClassName, project)
    }

    private fun getChildClassesRecursive(descendants: MutableList<String>, className: String, project: Project): List<String> {
        if (DumbService.isDumb(project)) {
            throw IndexNotReadyRuntimeException()
        }
        for (classDeclarationElement in get(className, project)) {
            val currentClassName = classDeclarationElement.classNameString
            if (descendants.contains(currentClassName)) {
                continue
            }
            descendants.add(currentClassName)
            getChildClassesRecursive(descendants, currentClassName, project)
        }
        return descendants
    }

    companion object {
        val instance = ObjJClassInheritanceIndex()
        private val KEY = IndexKeyUtil.createIndexKey<String, ObjJClassDeclarationElement>(ObjJClassInheritanceIndex::class.java)
        private val VERSION = 1
    }
}
